conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.runtime;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.conjure.java.api.errors.SerializableError;
+import com.palantir.conjure.java.serialization.ObjectMappers;
+import com.palantir.conjure.java.undertow.lib.Endpoint;
+import com.palantir.conjure.java.undertow.lib.ReturnValueWriter;
+import com.palantir.conjure.java.undertow.lib.UndertowRuntime;
+import io.undertow.Undertow;
+import io.undertow.server.HttpHandler;
+import io.undertow.util.Headers;
+import io.undertow.util.HttpString;
+import io.undertow.util.Methods;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.OptionalInt;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class AsyncRequestProcessingTest {
+
+    private static final ObjectMapper clientMapper = ObjectMappers.newClientObjectMapper();
+
+    private Undertow server;
+    private ListeningScheduledExecutorService scheduler;
+
+    @Before
+    public void before() {
+        this.scheduler = MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor());
+        UndertowRuntime runtime = ConjureUndertowRuntime.builder()
+                // Use a 1 second timeout so we can wait for the timeout to be hit without tests running too long.
+                .asyncTimeout(Duration.ofSeconds(1))
+                .build();
+        ReturnValueWriter<String> writer = (returnValue, exchange) -> {
+            exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/plain");
+            exchange.getResponseSender().send(returnValue);
+        };
+        HttpHandler httpHandler = exchange -> {
```

Comment:
I'd be interested in seeing the API that a service author would interact with. (I assume they won't get to see the exchange)

Reply:
ignore me, saw the thing down below